File: codeProjects/DemoTools/DemoTools/Counter.cpp
#include "Counter.h"

using namespace DemoTools;

Counter::Counter() : counter(0) {}

void Counter::Reset()
{
	counter = 0;
}

int Counter::Get() const
{
	return counter;
}

void Counter::Increase()
{
	counter++;
}

Counter & DemoTools::Counter::operator+(int value)
{
	counter += value;
	return *this;
}

Counter & DemoTools::Counter::operator+(const Counter & other)
{
	counter += other.counter;
	return *this;
}


Counter & DemoTools::MainCounter() {
	static Counter s;
	return s;
}

File: codeProjects/DemoTools/PCLDemoTools/Counter.cs
using System;
using System.Collections.Generic;
using System.Text;
using DemoTools;

namespace PCLDemoTools
{
    class Counter : DemoTools.Counter, DemoTools.ICounter
    {
        public int Value => Get();

        public void Add(ICounter other)
        {
            base.Add(other as Counter);
        }

        void ICounter.Add(int value)
        {
            base.Add(value);
        }
    }

    class MainCounter : DemoTools.ICounter
    {
        public void Reset()
        {
            DemoTools.DemoTools.MainCounter().Reset();
        }

        public void Add(int value)
        {
            DemoTools.DemoTools.MainCounter().Add(value);
        }

        public void Add(ICounter other)
        {
            DemoTools.DemoTools.MainCounter().Add(other.Value);
        }

        public int Value { get => DemoTools.DemoTools.MainCounter().Get(); }
    }
}
